src/sinks/tap.rs

```diff
@@ -0,0 +1,251 @@
+use super::util::StreamSink;
+use crate::{
+    buffers::Acker,
+    config::{Config, DataType, SinkConfig, SinkContext, SinkOuter},
+    event::{Event, LogEvent},
+};
+use futures::{future, stream::BoxStream, FutureExt, StreamExt};
+use indexmap::IndexMap;
+use parking_lot::RwLock;
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, sync::Arc};
+use tokio::sync::broadcast;
+use uuid::Uuid;
+
+type Sender = broadcast::Sender<LogEvent>;
+type Receiver = broadcast::Receiver<LogEvent>;
+
+#[derive(Debug)]
+pub struct TapSink {
+    tx: Sender,
+    acker: Acker,
+}
+
+/// Tap sink is used to 'tap' into events received by upstream components, and broadcast
+/// them to subscribers. Typically, this is used to expose events to the API, but is general
+/// purpose enough that it could technically be used with other mechanisms. This sink is
+/// not added to inventory; it's not intended to be user configurable.
+impl TapSink {
+    pub fn new(tx: Sender, acker: Acker) -> Self {
+        Self { tx, acker }
+    }
+}
+
+#[async_trait::async_trait]
+impl StreamSink for TapSink {
+    async fn run(&mut self, mut input: BoxStream<'_, Event>) -> Result<(), ()> {
+        while let Some(event) = input.next().await {
+            if let Event::Log(event) = event {
+                // This can suffer from TOC/TOU, but the risk is minimal as the purpose
+                // here is solely to reduce expense.
+                if self.tx.receiver_count() > 0 {
+                    let _ = self.tx.send(event.clone());
+                }
+            }
+            self.acker.ack(1);
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+/// Since 'tap' isn't user configurable, we're using `TapConfig` as middleware to satisfy the
+/// `SinkConfig` trait, and pass along the relaying `Sender` to the eventual sink. This is wrapped
+/// in a `RwLock` for interior mutability, since `SinkConfig.build` requires an immutability borrow.
+/// The `Option<Sender>` wrapper is used to replace the value with `None` when taken.
+pub struct TapConfig {
+    #[serde(skip_deserializing, skip_serializing, default = "default_locked_tx")]
+    locked_tx: RwLock<Option<Sender>>,
+}
+
+fn default_locked_tx() -> RwLock<Option<Sender>> {
+    RwLock::new(None)
+}
+
+impl TapConfig {
+    fn new(tx: Sender) -> Self {
+        Self {
+            locked_tx: RwLock::new(Some(tx)),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "tap")]
+impl SinkConfig for TapConfig {
+    async fn build(
+        &self,
+        cx: SinkContext,
+    ) -> crate::Result<(super::VectorSink, super::Healthcheck)> {
+        let mut lock = self.locked_tx.write();
```

Comment:
Using the blocking mutex in the async context will lead to reactor locking. I suggest using async-aware mutexes - they'll park the task while waiting, rather than blocking the thread.

Reply:
Do you have a suggested lib/lock that you'd recommend in this scenario? I saw https://lib.rs/crates/future-parking_lot, but have no experience with it.

Reply:
Perhaps https://docs.rs/tokio/1.2.0/tokio/sync/struct.RwLock.html ?

Reply:
(Or at least the current tokio version equivalent of it.)

Reply:
Yep, tokio mutexes are good! Also, https://docs.rs/futures/0.3.12/futures/lock/struct.Mutex.html, and a potentially more interesting https://docs.rs/futures/0.3.12/futures/lock/struct.BiLock.html - it's more lightweight than the usual mutex, but it is unconventional.
There's also an https://docs.rs/arc-swap/1.2.0/arc_swap/ - it's also doesn't block, so it's fine to use within async context.